model/tasks/ChangePermissionsTask.php

```diff
@@ -43,32 +45,114 @@
  */
 class ChangePermissionsTask extends AbstractAction implements TaskAwareInterface, JsonSerializable
 {
-    use ServiceManagerAwareTrait;
     use TaskAwareTrait;
     use OntologyAwareTrait;
 
-    public const PARAM_RECURSIVE = 'recursive';
     public const PARAM_RESOURCE = 'resource';
     public const PARAM_PRIVILEGES = 'privileges';
+    public const PARAM_RECURSIVE = 'recursive';
+    public const PARAM_NESTED_RESOURCES = 'nested_resources';
+    public const PARAM_REQUEST_ROOT = 'request_root';
+
+    private const MANDATORY_PARAMS = [
+        self::PARAM_RESOURCE,
+        self::PARAM_PRIVILEGES
+    ];
 
     public function __invoke($params = []): Report
     {
         $this->validateParams($params);
+
         try {
-            $service = $this->getPermissionService();
-            $service->savePermissions(
-                (bool)$params[self::PARAM_RECURSIVE],
+            return $this->doHandle(
                 $this->getClass($params[self::PARAM_RESOURCE]),
-                $params[self::PARAM_PRIVILEGES]
+                ($params[self::PARAM_REQUEST_ROOT] ?? $params[self::PARAM_RESOURCE]),
+                (array) $params[self::PARAM_PRIVILEGES],
+                (bool) ($params[self::PARAM_RECURSIVE] ?? false),
+                (bool) ($params[self::PARAM_NESTED_RESOURCES] ?? false)
             );
-            $result = Report::createSuccess('Permissions saved');
-        } catch (Exception $exception) {
-            $errMessage = sprintf('Saving permissions failed: %s', $exception->getMessage());
+        } catch (Exception $e) {
+            $errMessage = sprintf('Saving permissions failed: %s', $e->getMessage());
+
             $this->getLogger()->error($errMessage);
-            $result = Report::createFailure($errMessage);
+            return Report::createFailure($errMessage);
         }
+    }
+
+    private function doHandle(
+        core_kernel_classes_Class $root,
+        string $requestRoot,
+        array $privileges,
+        bool $isRecursive,
+        bool $withNestedResources
+    ): Report {
+        if ($withNestedResources) {
+            $message = sprintf(
+                "Permissions saved for resources under subclass %s [%s]",
+                $this->formatLabel($root),
+                $root->getUri()
+            );
 
-        return $result;
+            $command = new ChangePermissionsCommand($root, $requestRoot, $privileges);
+            $command->withNestedResources();
+
+            $this->getPermissionService()->applyPermissions($command);
+        } elseif ($isRecursive) {
+            $message = 'Starting recursive permissions update';
+
+            $this->createSubtasksForClasses(
+                array_merge(
+                    [$root],
+                    $root->getSubClasses(true) // recursive, NOT including the root
```

Comment:
TODO
- [ ] Should be obvious: `getSubclasses` returns subclasses but not `$root` itself, remove the comment